### PR TITLE
feat(2437): Support read-only SCM [4]

### PR DESCRIPTION
**Pull request by tkyi**
## Context

Should support read-only SCM.

## Objective

This PR:
- adds validation for `readOnly` config in SCM options
- takes in proper `cloneType` for `getCheckoutCommand()`
- fixes `addPRComment()` and other PR comment methods

## References

Related https://github.com/screwdriver-cd/screwdriver/issues/2437

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -790,29 +871,57 @@ class GitlabScm extends Scm {
...
+
+            if (botComment) {
+                try {
+                    const pullRequestComment = await this.editPrComment(
+                        botComment.id, repoId, prNum, comment);
+
+                    return {
+                        commentId: `${pullRequestComment.body.id}`,
+                        createTime: `${pullRequestComment.body.created_at}`,
+                        username: pullRequestComment.body.author.username
```
Comment (pritamstyz4ever) [inline index.js:891]: Update after Line 855 and check for undefined.

File: index.js [hunk 2]
```diff
@@ -790,29 +871,57 @@ class GitlabScm extends Scm {
...
+                method: 'POST',
+                auth: {
+                    bearer: this.config.commentUserToken
+                },
+                url: `${this.config.gitlabProtocol}://${this.config.gitlabHost}/api/v4` +
+                     `/projects/${repoId}/merge_requests/${prNum}/notes`,
+                qs: {
+                    body: comment
+                }
+            });
```
Comment (pritamstyz4ever) [inline index.js:913]: if it is convenient, refactor the api calls to a single func with url as param

**Review by pritamstyz4ever — CHANGES_REQUESTED** [review]

File: index.js (hunk 2 above)
Comment (tkyi) [reply index.js:913]: will update at a different time

File: index.js [hunk 3]
```diff
@@ -779,6 +799,67 @@ class GitlabScm extends Scm {
...
+                    bearer: this.config.commentUserToken // need to use a token with public_repo permissions
+                },
+                url: `${this.config.gitlabProtocol}://${this.config.gitlabHost}/api/v4` +
+                     `/projects/${repoId}/merge_requests/${prNum}/notes/${commentId}`,
+                qs: {
+                    body: comment
+                }
+            });
+
+            return pullRequestComment;
```
[the remark this answers is not in the document]
Comment (tkyi) [reply index.js:855]: this change is causing some issues with tests, so leaving as is

File: index.js [hunk 4]
```diff
@@ -779,6 +799,67 @@ class GitlabScm extends Scm {
...
+
+            return null;
+        }
+    }
+
+    /**
+     * Edit a particular comment in the PR
+     * @async  editPrComment
+     * @param  {Integer}  commentId         The id of the particular comment to be edited
+     * @param  {Object}   scmInfo           The information regarding SCM like repo, owner
```
[the remark this answers is not in the document]
Comment (tkyi) [reply index.js:836]: updated

File: index.js [hunk 5]
```diff
@@ -779,6 +799,67 @@ class GitlabScm extends Scm {
...
+                auth: {
+                    bearer: this.config.commentUserToken
+                },
+                url: `${this.config.gitlabProtocol}://${this.config.gitlabHost}/api/v4` +
+                     `/projects/${repoId}/merge_requests/${prNum}/notes`
+            });
+
+            return { comments: prComments.body };
+        } catch (err) {
+            logger.warn('Failed to fetch PR comments: ', prComments.body.message
```
[the remark this answers is not in the document]
Comment (tkyi) [reply index.js:825]: updated

**Review by pritamstyz4ever — APPROVED** [review]

**Pull request merged**